crates/proof/proof/src/l1/blob_provider.rs

```diff
@@ -91,3 +96,113 @@ impl<T: CommsClient + Sync + Send> BlobProvider for OracleBlobProvider<T> {
         Ok(blobs)
     }
 }
+
+/// The 4096th bit-reversed roots of unity used in EIP-4844 as predefined evaluation points.
+///
+/// See `generate_roots_of_unity` for details on how these roots of unity are generated.
+pub static ROOTS_OF_UNITY: Lazy<[Fr; FIELD_ELEMENTS_PER_BLOB as usize]> =
+    Lazy::new(generate_roots_of_unity);
+
+/// Generates the 4096th bit-reversed roots of unity used in EIP-4844 as predefined evaluation
+/// points. To compute the field element at index i in a blob, the blob polynomial is evaluated at
+/// the i'th root of unity. Based on go-kzg-4844: <https://github.com/crate-crypto/go-kzg-4844/blob/8bcf6163d3987313a3194595cf1f33fd45d7301a/internal/kzg/domain.go#L44-L98>
+/// Also, see the consensus specs:
+///   - compute_roots_of_unity <https://github.com/ethereum/consensus-specs/blob/bf09edef17e2900258f7e37631e9452941c26e86/specs/deneb/polynomial-commitments.md#compute_roots_of_unity>
+///   - bit-reversal permutation: <https://github.com/ethereum/consensus-specs/blob/bf09edef17e2900258f7e37631e9452941c26e86/specs/deneb/polynomial-commitments.md#bit-reversal-permutation>
+fn generate_roots_of_unity() -> [Fr; FIELD_ELEMENTS_PER_BLOB as usize] {
+    const MAX_ORDER_ROOT: u64 = 32;
+
+    let mut roots_of_unity = [Fr::ZERO; FIELD_ELEMENTS_PER_BLOB as usize];
+
+    // Generator of the largest 2-adic subgroup of order 2^32.
+    let root_of_unity = Fr::new(
+        BigInteger256::from_str(
+            "10238227357739495823651030575849232062558860180284477541189508159991286009131",
+        )
+        .expect("Failed to initialize root of unity"),
+    );
+
+    // Find generator subgroup of order x.
+    // This can be constructed by powering a generator of the largest 2-adic subgroup of order 2^32
+    // by an exponent of (2^32)/x, provided x is <= 2^32.
+    let log_x = FIELD_ELEMENTS_PER_BLOB.trailing_zeros() as u64;
+    let expo = 1u64 << (MAX_ORDER_ROOT - log_x);
+
+    // Generator has order x now
+    let generator = root_of_unity.pow([expo]);
+
+    // Compute all relevant roots of unity, i.e. the multiplicative subgroup of size x
+    let mut current = Fr::ONE;
+    (0..4096).for_each(|i| {
+        roots_of_unity[i] = current;
+        current *= generator;
+    });
+
+    let shift_correction = 64 - FIELD_ELEMENTS_PER_BLOB.trailing_zeros();
+    (0..FIELD_ELEMENTS_PER_BLOB).for_each(|i| {
+        // Find index irev, such that i and irev get swapped
+        let irev = i.reverse_bits() >> shift_correction;
+        if irev > i {
+            roots_of_unity.swap(i as usize, irev as usize);
+        }
+    });
+
+    roots_of_unity
+}
+
+#[cfg(test)]
+mod test {
+    use super::ROOTS_OF_UNITY;
+    use alloy_eips::eip4844::{FIELD_ELEMENTS_PER_BLOB, env_settings::EnvKzgSettings};
+    use ark_ff::{BigInteger, PrimeField};
+    use c_kzg::{BYTES_PER_BLOB, Blob, Bytes32, Bytes48};
+    use rand::Rng;
+
+    #[test]
+    fn test_slow_roots_of_unity() {
```

Comment:
Note to reader - this test takes a long time (~1m30s) to run. I've excluded it from the default `test` directive, as the functionality here is not subject to frequent change.

Reply:
is there a faster smoke test we could add alongside this one?

Reply:
I could maybe pull in KZG-rs, which is parallel to do this. Didn't want to add a dep that heavy just for testing, though.

Perhaps could use a fully 0'd out blob? That would likely be quite a bit faster.

Reply:
er actually, I'll just try to parallelize this rq. 1 sec.

Reply:
Parallelizing it brought it down to 9-10s on my machine. Probably good enough to include in CI.

Reply:
Nice. i was thinking of checking a single field element as a fast, but non-comprehensive test. Rayon power is even better.